File: frontend/schoolio-web-app/src/components/Folder/FolderDetailedDTO.ts
import { FolderDTO } from './FolderDTO';
import {NoteDTO} from "../Note/NoteDTO";

export class FolderDetailedDTO {
    id: number;
    name: string;
    parentFolder: FolderDTO | null;
    subFolders: FolderDTO[];
    notes: NoteDTO[];

    constructor(data: any) {
        this.id = data.id;
        this.name = data.name;
        this.parentFolder = data?.parentFolder ? new FolderDTO(data.parentFolder) : null;
        this.subFolders = data.subFolders.map((subFolder: any) => new FolderDTO(subFolder));
        this.notes = data.notes.map((note: any) => new NoteDTO(note));
    }

    isEmpty(): boolean {
        return this.subFolders.length == 0 && this.notes.length == 0;
    }
}

File: api/schoolio-api/src/main/java/org/example/schoolioapi/DTO/Folder/FolderDetailedDTO.java
package org.example.schoolioapi.DTO.Folder;

import lombok.Builder;
import org.example.schoolioapi.DTO.Note.NoteDTO;
import org.example.schoolioapi.domain.Folder;

import java.util.List;

@Builder
public record FolderDetailedDTO(
        Long id,
        String name,
        FolderDTO parentFolder,
        List<FolderDTO> subFolders,
        List<NoteDTO> notes

) {
    public static FolderDetailedDTO from(Folder folder) {
        if (folder == null) return null;
        return FolderDetailedDTO.builder()
                .id(folder.getId())
                .name(folder.getName())
                .parentFolder(FolderDTO.from(folder.getParent()))
                .subFolders(folder.getSubFolders().stream().map(FolderDTO::from).toList())
                .notes(folder.getNotes().stream().map(NoteDTO::from).toList())
                .build();
    }
}
